packages/xstate-immer/src/index.ts

```diff
@@ -2,51 +2,62 @@ import {
   EventObject,
   ActionObject,
   AssignAction,
-  assign as xstateAssign
+  assign as xstateAssign,
+  AssignMeta
 } from 'xstate';
-import { produce, Draft, produceWithPatches, applyPatches } from 'immer';
+import { produce, Draft } from 'immer';
 
 export type ImmerAssigner<TContext, TEvent extends EventObject> = (
   context: Draft<TContext>,
-  event: TEvent
+  event: TEvent,
+  meta: AssignMeta<TContext, TEvent>
 ) => void;
 
 export interface ImmerAssignAction<TContext, TEvent extends EventObject>
   extends ActionObject<TContext, TEvent> {
   assignment: ImmerAssigner<TContext, TEvent>;
 }
 
-export function assign<TContext, TEvent extends EventObject = EventObject>(
-  assignment: ImmerAssigner<TContext, TEvent>
+function immerAssign<TContext, TEvent extends EventObject = EventObject>(
+  recipe: ImmerAssigner<TContext, TEvent>
 ): AssignAction<TContext, TEvent> {
-  // @ts-ignore (possibly infinite TS bug)
-  return xstateAssign((context, event) => {
-    return produce(context, (draft) => void assignment(draft, event));
+  return xstateAssign((context, event, meta) => {
+    return produce(context, (draft) => void recipe(draft, event, meta));
   });
 }
 
-interface PatchEventObject extends EventObject {
-  patches: ReturnType<typeof produceWithPatches>;
+export { immerAssign as assign };
+
+export interface ImmerUpdateEvent<
+  TType extends string = string,
+  TInput = unknown
+> {
+  type: TType;
+  input: TInput;
 }
 
-export function patchEvent<TContext>(
-  type: string,
-  context: TContext,
-  recipe: (draftContext: TContext) => void
-): PatchEventObject {
-  return {
-    type,
-    patches: produceWithPatches(context, recipe)
-  };
+export interface ImmerUpdater<TContext, TEvent extends ImmerUpdateEvent> {
+  update: (input: TEvent['input']) => TEvent;
+  action: AssignAction<TContext, TEvent>;
+  type: TEvent['type'];
 }
 
-export function assignPatch<
-  TContext,
-  TEvent extends PatchEventObject
->(): AssignAction<TContext, TEvent> {
-  return xstateAssign((context, event) => {
-    const [, patches] = event.patches;
+export function createUpdater<TContext, TEvent extends ImmerUpdateEvent>(
+  type: TEvent['type'],
+  recipe: (ctx: Draft<TContext>, input: TEvent['input']) => void
```

Comment:
wondering - shouldn't this have matching signature to `assign` and `immerAssign`?

Reply:
I thought that too but that didn't make sense initially - the created event only consists of the type and the `input` (which is essentially the event payload) itself.

Doing otherwise would force the user to always destructure:

```js
createUpdater('UPDATE_SOMETHING', (ctx, { input }) => {
 // ...
});
```

But now I see use-cases where having something besides the `input` might be beneficial.